include/pisa/topk_queue.hpp

```diff
@@ -69,17 +76,25 @@ struct topk_queue {
 
     /// Sorts the results in the heap container in the descending score order.
     ///
-    /// After calling this function, the heap should be no longer modified, as
-    /// the heap order will not be preserved.
+    /// If multiple entries have equal score, they are sorted by document ID. Notice that this only
+    /// happens in the finalization step; due to performance considerations, inserting is done with
+    /// score-only order. After calling this function, the heap should be no longer modified, as the
+    /// heap order will not be preserved.
     void finalize()
     {
-        std::sort_heap(m_q.begin(), m_q.end(), min_heap_order);
-        size_t size = std::lower_bound(
-                          m_q.begin(),
-                          m_q.end(),
-                          0,
-                          [](std::pair<Score, DocId> l, Score r) { return l.first > r; })
-            - m_q.begin();
+        auto sort_order = [](auto const& lhs, auto const& rhs) {
+            if (lhs.first == rhs.first) {
+                return lhs.second < rhs.second;
+            }
+            return lhs.first > rhs.first;
+        };
+        // We have to do a full sort because it is not the exact same ordering as when pushing
+        // elements to the heap.
+        std::sort(m_q.begin(), m_q.end(), sort_order);
```

Comment:
If we are switching to `sort` then why don't we just use `stable_sort` on the scores such that we maintain as second sorting condition the insertion time?

Reply:
I'm not sure `stable_sort` will preserve that. I actually haven't analyzed this in detail, but are you saying that given our implementation, we can be sure that the more recently inserted entry is always above in the heap or at the same level but on the left side?

Reply:
The issue is that `heap_sort` creates some issues with reproducibility since two documents with equal score have no guarantees on the order they will appear in the list of results.
Stable sort will keep the relative order of elements with equivalent values, fixing the reproducibility issue.

I prefer this solution because:
(1) it is easier and less error prone
(2) it resolves the issue with docID ordering not maintained while inserting and only done at the end, which could cause the problem described in the code comment.

Reply:
So what you're suggesting is to change the semantic: instead of ensuring that within the top-k results ties are broken by the docID order, you're proposing to not give any such guarantee, but rather say that the order of elements of with the same score is undefined, but deterministic, i.e., reproducible across multiple runs.

Do I understand correctly?

Reply:
> it resolves the issue with docID ordering not maintained while inserting and only done at the end, which could cause the problem described in the code comment

You'll have to be more explicit here, not sure which issue we're resolving. From my understanding, you're solution simply changes the semantic of what we do, but maybe I'm missing something, so if you could be a bit more clear about what you mean and how it solves it, that would help.